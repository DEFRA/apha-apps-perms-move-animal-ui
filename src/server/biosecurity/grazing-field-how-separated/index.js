import { Page } from '../../common/model/page/page-model.js'

// using Page to not make whole section invalid due to the page not being implemented yet
// make QuestionPage when implementing
export class GrazingFieldHowSeparatedPage extends Page {
  urlPath = '/biosecurity/grazing-field-how-separated'
  sectionKey = 'biosecurity'
  question = ''
  questionKey = 'grazingFieldHowSeparated'
}

export const grazingFieldHowSeparatedPage = new GrazingFieldHowSeparatedPage()
