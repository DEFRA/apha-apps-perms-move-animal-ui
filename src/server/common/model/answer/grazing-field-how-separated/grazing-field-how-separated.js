import { TextAreaAnswer } from '../text-area/text-area.js'

/** @import {TextAreaConfig} from '../text-area/text-area.js' */

/**
 * @typedef {{ grazingFieldHowSeparated: string }} GrazingFieldHowSeparatedPayload
 */

/**
 * @augments {TextAreaAnswer<GrazingFieldHowSeparatedPayload>}
 */
export class GrazingFieldHowSeparatedAnswer extends TextAreaAnswer {
  /** @type {TextAreaConfig} */
  static config = {
    payloadKey: 'grazingFieldHowSeparated',
    hint: 'Include the types of fencing you use and distances between them',
    rows: 8,
    validation: {
      maxLength: {
        value: 5000,
        message: 'Your answer must be no longer than 5000 characters'
      },
      empty: {
        message:
          'Enter information about how this grazing field is separated from the resident herd'
      }
    }
  }
}
