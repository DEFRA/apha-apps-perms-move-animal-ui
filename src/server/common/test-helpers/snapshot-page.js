import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
