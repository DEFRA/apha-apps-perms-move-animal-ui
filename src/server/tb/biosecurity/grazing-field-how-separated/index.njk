{% from "govuk/components/button/macro.njk" import govukButton %}
{% extends 'layouts/questions.njk' %}
{% block heading %}
  <h1 class="govuk-heading-l">Separating the incoming cattle</h1>
{% endblock %}
{% block beforequestion %}
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <p>
        We need to know what measures are in place to reduce the risk of
        spreading TB to the incoming animals from the resident herd.
      </p>
      <p>You should include:</p>
      <ul class="govuk-list govuk-list--bullet">
        <li>
          minimum distances between incoming animals and the resident herd
        </li>
        <li>the types of fencing you use</li>
        <li>how grazing fields are kept separate</li>
      </ul>
    </div>
  </div>
{% endblock %}
