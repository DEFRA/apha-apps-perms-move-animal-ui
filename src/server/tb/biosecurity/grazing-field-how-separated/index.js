import { QuestionPageController } from '../../../common/controller/question-page-controller/question-page-controller.js'
import { GrazingFieldHowSeparatedAnswer } from '../../../common/model/answer/grazing-field-how-separated/grazing-field-how-separated.js'
import { QuestionPage } from '../../../common/model/page/question-page-model.js'
import { lastGrazedPage } from '../last-grazed/index.js'

export class GrazingFieldHowSeparatedPage extends QuestionPage {
  urlPath = '/biosecurity/grazing-field-how-separated'
  sectionKey = 'biosecurity'
  question = 'How will you separate the incoming cattle from the resident herd?'
  questionKey = 'grazingFieldHowSeparated'
  Answer = GrazingFieldHowSeparatedAnswer
  view = 'tb/biosecurity/grazing-field-how-separated/index.njk'

  nextPage() {
    return lastGrazedPage
  }
}

export const grazingFieldHowSeparatedPage = new GrazingFieldHowSeparatedPage()

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const grazingFieldHowSeparated = new QuestionPageController(
  grazingFieldHowSeparatedPage
).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
