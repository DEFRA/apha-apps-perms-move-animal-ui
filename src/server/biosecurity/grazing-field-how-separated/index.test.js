import { GrazingFieldHowSeparatedAnswer } from '../../common/model/answer/grazing-field-how-separated/grazing-field-how-separated.js'
import { describePageSnapshot } from '../../common/test-helpers/snapshot-page.js'
import { buildingsAnySharedPage } from '../buildings-any-shared/index.js'
import {
  grazingFieldHowSeparatedPage,
  GrazingFieldHowSeparatedPage
} from './index.js'

const sectionKey = 'biosecurity'
const question = 'How is this grazing field separated from the resident herd?'
const questionKey = 'grazingFieldHowSeparated'
const view = 'common/model/page/question-page.njk'
const pageUrl = '/biosecurity/grazing-field-how-separated'

describe('GrazingFieldHowSeparatedPage', () => {
  const page = new GrazingFieldHowSeparatedPage()

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(GrazingFieldHowSeparatedAnswer)
  })

  it('nextPage should return any shared buildings page', () => {
    const nextPage = page.nextPage()
    expect(nextPage).toBe(buildingsAnySharedPage)
  })

  it('should export page', () => {
    expect(grazingFieldHowSeparatedPage).toBeInstanceOf(
      GrazingFieldHowSeparatedPage
    )
  })

  describePageSnapshot({
    describes: 'grazingFieldHowSeparatedPage.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
