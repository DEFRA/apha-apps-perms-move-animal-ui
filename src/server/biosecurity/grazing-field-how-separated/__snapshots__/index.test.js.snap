// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrazingFieldHowSeparatedPage grazingFieldHowSeparatedPage.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<fieldset class="govuk-fieldset">
  <div class="govuk-form-group">
  <h1 class="govuk-label-wrapper">
    <label class="govuk-label govuk-label--l" for="grazingFieldHowSeparated">
      How is this grazing field separated from the resident herd?
    </label>
  </h1>
  <div id="grazingFieldHowSeparated-hint" class="govuk-hint">
    Include the types of fencing you use and distances between them
  </div>
  <textarea class="govuk-textarea" id="grazingFieldHowSeparated" name="grazingFieldHowSeparated" rows="8" aria-describedby="grazingFieldHowSeparated-hint"></textarea>
</div>


</fieldset>
            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
